core/jvm/src/test/scala/fs2/CompressSpec.scala

```diff
@@ -73,93 +85,426 @@ class CompressSpec extends Fs2Spec {
           val actualInflated = Stream
             .chunk(Chunk.bytes(actual))
             .covary[Fallible]
+            .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
             .through(inflate(nowrap = nowrap))
             .compile
             .toVector
           actualInflated should equal(Right(expectedInflated))
         }
 
-        expectEqual(actualDeflated.toArray, expectedDeflated.toArray)
-        expectEqual(expectedDeflated.toArray, actualDeflated.toArray)
+        expectEqual(actualDeflated.toArray, expectedDeflated)
+        expectEqual(expectedDeflated, actualDeflated.toArray)
     }
 
     "deflate |> inflate ~= id" in forAll { s: Stream[Pure, Byte] =>
       s.covary[IO]
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
         .through(compress.deflate())
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
+        .through(compress.inflate())
+        .compile
+        .toVector
+        .asserting(_ shouldBe s.toVector)
+    }
+
+    "classic.deflate |> inflate ~= id" in forAll { s: Stream[Pure, Byte] =>
+      s.covary[IO]
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
+        .through(classic.deflate())
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
         .through(compress.inflate())
         .compile
         .toVector
         .asserting(_ shouldBe s.toVector)
     }
 
+    "deflate |> classic.inflate ~= id" in forAll { s: Stream[Pure, Byte] =>
+      s.covary[IO]
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
+        .through(compress.deflate())
+        .rechunkRandomlyWithSeed(0.1, 2)(System.nanoTime())
+        .through(classic.inflate())
+        .compile
+        .toVector
+        .asserting(_ shouldBe s.toVector)
+    }
+
+    "deflate performance" - {
```

Comment:
This really isn't a very good way to measure performance of *either* implementation, just due to the way the JVM works. JMH is a considerably better tool for this, and I believe there's already a **benchmarks** subproject.

Reply:
Agreed, this is not a proper benchmark. I added these for a rough comparison and sanity check. The real before and after performance metrics were gathered by profiling with Oracle Java Mission Control's Flight Recorder. At one point compress.inflate was on of the hottest areas of code in a application heavy on streaming compressed json from networked sources.

Would you prefer I just remove these false performance tests along with the classic implementation?

Reply:
> Would you prefer I just remove these false performance tests along with the classic implementation?

I think that would definitely be better. Also I'm curious as to your JFR results!

Reply:
![fs2_inflate](https://user-images.githubusercontent.com/11074540/74085648-33dc6e00-4a73-11ea-9aa6-3fecbe927ffd.png)

Of course. Attached is a screenshot of the JFR hot methods trace that started this effort, namely attempting to avoid the mutable ArrayBuffer used to accumulate the bytes produced by Inflater. The context was streaming multi-megabyte and multi-gigabyte compressed json documents.

In the spirit of fs2 and to draw on all your insight and improvements it would be good to feedback these enhancements that are already in use in this area.

While this PR relates to efficiently chunking the compressed byte stream, a bigger challenge was efficiently recombining the chunked byte stream with constant time complexity. If there's interest, I can raise a PR for BytesVector (signature below) with constant time append and Chunk implementations. 

`final case class Chunk.BytesVector(vector: Vector[Chunk.Bytes]) extends Chunk[Byte] with KnownElementType[Byte]`

The combination of these enhancements made it possible to build efficient streaming json parsing from wire to instance for very large objects. Chunking byte streams has historically been the domain of C-style array-length-offset arithmetic and mutable arrays, which is difficult to compete with from the perspective of raw performance. But in the context of integrating raw byte streams with complex and useful applications, like can be done by building streams with fs2, a better performing overall result can be achieved.

Reply:
Have you looked at `Chunk.Queue[Byte]`? It's a wrapper around a `scala.collection.immutable.Queue[Chunk[A]]` with efficient conversion to a single array backed chunk.

Reply:
Another option is to use `Chunk.ByteVector` and just append `ByteVector`s as they are received.

Reply:
> The combination of these enhancements made it possible to build efficient streaming json parsing from wire to instance for very large objects. Chunking byte streams has historically been the domain of C-style array-length-offset arithmetic and mutable arrays, which is difficult to compete with from the perspective of raw performance. But in the context of integrating raw byte streams with complex and useful applications, like can be done by building streams with fs2, a better performing overall result can be achieved.
 
Worth noting that this is basically *directly* [in my wheelhouse](https://github.com/slamdata/tectonic). Tectonic started as a fork of Jawn, and basically uses the same approach: byte chunks are "absorbed" into the main contiguous buffer, which is then parsed until some partial-failure is hit. Every partial-success (e.g. successfully parsing a key, a `:`, a `}`, etc) is a stateful safepoint within the parser, advancing the cursor and persisting (in registers) all the information required to resume parsing from that offset. Thus, when we run off the end with a partial result, we can just hard-fail with an exception (which is cached). The exception effectively just routes control flow back to the stream `Pull`, asynchronously suspending the parser until the next chunk comes in.

All parsing operations are done against a straight, contiguous byte array (which grows up to 1 MB). The downside to this approach is that it isn't possible to avoid paging the bytes fully into memory (e.g. for mmap'd files and such), but in practice `ByteBuffer` is insanely slow anyway, even when you play tricks to get monomorphic access.

Anyway, performance in practice saturates the IO bus, even when performing relatively complex transformations on the *results* of the parse.

Reply:
Let's say the json document being streamed is 2GB. It's not ideal to materialise 65k x 32KB chunks into a single array and even if that is only done once it will still require at least 4GB of memory - so it's not ideal to use Chunk.Queue[Byte] in this context. Chunk.ByteVectorChunk's ByteVector documents logarithmic time complexity for most operations, but I'll run a comparison when I get a chance.

Performing streaming tokenization for streaming parsing of the large json documents doesn't require materialising the entire document into one array or one String as only localised concatenation is required for tokens that span sequential chunks, which can be achieved with a view (similar to ByteVector's view approach) that acts as a sliding window over the localised scope of the streaming tokenizer. The single class, Vector-backed implementation of Chunk[Byte] I mentioned is nowhere near the scope of what I read Tectonic to be, but I'll still be happy to share it so it can be improved or replaced.

Reply:
> Let's say the json document being streamed is 2GB. It's not ideal to materialise 65k x 32KB chunks into a single array and even if that is only done once it will still require at least 4GB of memory

Tectonic only does it up to a maximum bound of 1 MB. It discards anything that it has already parsed, including fragments. In order to create a true overflow of the type you suggest, you would need a single *json string* which is 2 GB in size, which is quite rare. Basically, Tectonic is doing exactly the sliding window you're talking about, but with a raw array and width-sensitive to the exact token state.

> The single class, Vector-backed implementation of Chunk[Byte] I mentioned is nowhere near the scope of what I read Tectonic to be, but I'll still be happy to share it so it can be improved or replaced.

I agree. And more relevantly, Tectonic's internal byte paging machinery is quite complex and specialized. If what you've implemented is a more general mechanism that achieves the same thing with similar performance but without the manic complexity in Tectonic, that would be really wonderful.

Reply:
A good discussion, thanks. So, a single json string of 2GB is commonplace for my use case, hence this PR suggestion. I would have to go look what the maximum currently is, but I suspect I would fine two or three times that size. Why single json strings of that size is a different discussion entirely.

The chunk implementation that I found useful to support streaming tokenisation into streaming parsing of multi-gig json objecs is here: https://github.com/functional-streams-for-scala/fs2/pull/1789 - I would need to add a bunch of tests to ensure nothing is broken in this incarnation, so for now treat the PR as worthy of discussion only.

This BytesVector implementation enables tokenisation without array copying or concatenation because the tokens can be backed by the original byte chunks of the underlying stream, even when those tokens straddle byte chunks or span multiple byte chunks. If found it useful to constrain the implementation to always have an effective offset of zero and a total length equal to the sum of the length of the underlying Chunk.Byte instances. This reduced the cost complexity of of implementing the Chunk interface and ensured that only the visible chunks and their visible bytes for the required offset and length were retained and the out-of-scope chunks ignored. This was implemented by rewriting the leading and trailing Chunk.Byte to maintain only the in-scope bytes.

Once tokenised, parsing is just a matter of scanning the token and rewriting the escaped characters, depending on the token type - but you're way more familiar with Jawn, and as you've mentioned Tectonic, than I am. But, 99% of the time tokens do not contain escaped characters or multi-byte characters and are exactly the length of the parsed node, which is a good hint at the parsed node size. I imagine that with a UTF-16 encoding constraint and a String implementation that didn't instantiate and copy into a new array it would be possible to back the parsed nodes with the original byte chunks, but this would introduce new challenges. (Not that I haven't see java.lang.String rewritten and replaced for similar challenges before.)

Reply:
Thanks to the suggestion by @mpilquist, scodec.bits.ByteVector performed better for my use case. A rough, but reasonable performance comparison using a single 416MB json document with 38,831,305 tokens, was tokenised by the proposed Chunk.BytesVector in 11.894 seconds and by the recommended codec.bits.ByteVector in 7.871 seconds.

Reply:
So around 423 Mbps, and your average token size is 11.23 bytes. Very impressive.